app/engine/Rower.js

```diff
@@ -0,0 +1,357 @@
+'use strict'
+/*
+  Open Rowing Monitor, https://github.com/laberning/openrowingmonitor
+
+  The Rowing Engine models the physics of a real rowing boat.
+  It takes impulses from the flywheel of a rowing machine and estimates
+  parameters such as energy, stroke rates and movement.
+
+  This implementation uses concepts that are described here:
+  Physics of Rowing by Anu Dudhia: http://eodg.atm.ox.ac.uk/user/dudhia/rowing/physics
+  Also Dave Vernooy has some good explanations here: https://dvernooy.github.io/projects/ergware
+*/
+
+import loglevel from 'loglevel'
+import { createFlywheel } from './Flywheel.js'
+import { createCurveMetrics } from './utils/curveMetrics.js'
+
+const log = loglevel.getLogger('RowingEngine')
+
+function createRower (rowerSettings) {
+  const flywheel = createFlywheel(rowerSettings)
+  const sprocketRadius = rowerSettings.sprocketRadius / 100
+  const driveHandleForce = createCurveMetrics(2)
+  const driveHandleVelocity = createCurveMetrics(3)
+  const driveHandlePower = createCurveMetrics(1)
+  let _strokeState = 'WaitingForDrive'
+  let _totalNumberOfStrokes = -1.0
```

Comment:
What is the reason you are initializing the stroke count with -1? This means that you are not counting the first stroke.
Also when state goes to reset in the reset() function this is reinitialized as 0.

Reply:
Hahaha, that is a puzzling one isn't it?

First thing to realize is that some post processing apps completely ignore the total distance and total time fields and thus require a stroke 0: a stroke that defines the start position of the session (with distance 0, speed 0, starting HR, etc) to calculate a distance travelled between the first and last stroke.

Second thing to realize is that as soon as there is a transition from wait/recovery to a drive, a stroke is reported.

So what happens in practice is that the first currentDt's are reported as the flywheel starts spinning, thus the first drive is detected, and thus ORM concludes that there is a new stroke to be counted, which will be stroke 0, with no distance and speed. This will be reported to the workout recorder, which will record it. After completing the drive and recovery, stroke 1 is reported, etc..

The re-initialization is a bug, good catch!

Reply:
I am not sure I fully understand the reasoning but I accept that this is intentional. The reason for asking (apart from the noticed inconsistency on the initialization) is that I am trying to work out a standardized validation method for the stroke detection algorithm. For this I have chosen the number of stroke as control (that is what I can validate 100% from other sources) and I had discrepancies (that now I understand where they are coming from)

Reply:
Typically, you end up with one extra stroke (i.e. the stroke 0). If you terminate it based on fixed time or distance, you typically get two extra (the last stroke number is doubled, as typically you reach the limit mid-stroke).